src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -0,0 +1,141 @@
+from pip._vendor.packaging.utils import canonicalize_name
+
+from pip._internal.req.req_install import InstallRequirement
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+from .base import Candidate, Requirement, format_name
+
+if MYPY_CHECK_RUNNING:
+    from typing import (Optional, Sequence)
+
+    from pip._vendor.packaging.version import _BaseVersion
+
+    from pip._internal.index.package_finder import PackageFinder
+
+
+def make_requirement(install_req):
+    # type: (InstallRequirement) -> Requirement
+    if install_req.link:
+        if install_req.req and install_req.req.name:
+            return NamedRequirement(install_req)
+        else:
+            return UnnamedRequirement(install_req)
+    else:
+        return VersionedRequirement(install_req)
+
+
+class UnnamedRequirement(Requirement):
+    def __init__(self, req):
+        # type: (InstallRequirement) -> None
+        self._ireq = req
+        self._candidate = None  # type: Optional[Candidate]
+
+    @property
+    def name(self):
+        # type: () -> str
+        assert self._ireq.req is None or self._ireq.name is None, \
+            "Unnamed requirement has a name"
+        # TODO: Get the candidate and use its name...
+        return ""
```

Comment:
Are there cases where `name` wouldn’t be called? I can’t think of any. I am tempted to suggest always `NamedRequirement` instead, by getting the candidate *very* eagerly in `make_requirement` and assigning it to the `InstallRequirement`.

Reply:
I.e. something like

```python
def make_requirement(install_req):
    # type: (InstallRequirement) -> Requirement
    if not install_req.link:
        return VersionedRequirement(install_req)
    if not install_req.req:
        install_req.req = _parse_link_into_req(install_req.link)  # TODO: Implement me.
    named_req = NamedRequirement(install_req)
    if not install_req.req.name:
        candidate = _make_candidate(install_req)  # TODO: Implement me.
        named_req._candidate = candidate
    return named_req
```

Reply:
Agreed, but without a "proper" candidate object, I can't do this without the tests failing (see the test that currently asserts that a sdist has a name of ""). So I plan on addressing this in one of the follow up PRs (either the one that implements candidates, or more likely a further PR that tidies up all the links).

In hindsight, I'm not convinced that splitting up the original change from #7799 was worthwhile, it's created a lot of busy-work, and I don't think the resulting PRs are *that* much more reviewable. But adding the tests was (IMO) a definite benefit, so I'm not going to worry too much..